### DISCUSSION TITLE
Best practice for Provider depends on other providers

### DISCUSSION THREAD

**Discussion by XuanTung95** [rrousselGit/riverpod#1259] [Q&A]
Good day to you.
I've been using Riverpod for a while now. Everything is good but I'm confused about this use case. I want to know the community's opinion about the solution.

Case: I have 2 providers P1 and P2:
- P1 has data pulled from the server. The data can be changed over time as the user requests to pull the new value.
- P2 uses the data of P1 and has its own data to manage.
- When P1 gets new data, it should notify P2 about the change so P2 can update according to the incoming data.

The solutions I come up with:
- S1: Combine 2 P1 and P2 into one provider. This way the provider may be very large and hard to manage.
- S2: Combining provider states as described in the [doc](https://riverpod.dev/docs/concepts/combining_providers). I really don't like this because it creates a new state every time and the old state is lost.
- S3: P2 use ref.listen on P1. When P1 change, P2 call update method and pass new data of P1:
```
  static final provider = ChangeNotifierProvider((ref) {
    var ret = P2();
    ref.listen<P1>(P1.provider, (previous, next) {
      ret.updateCategories(next.getCategories());
    });
    var meta = ref.read(P1.provider);
    ret.updateCategories(meta.getCategories());
    return ret;
  });
```
This way should work I think. But it's verbose.

What do you think is the best way to handle this case?

**Comment by ghost** [comment]
I mainly use third solution. Even though it's verbose, and you might end up shooting your self in the foot(if you don't know what you're doing and don't consider possible edge cases), but it's way more customizable
side note: you might have reasons that I'm not aware of for writing it like this, but you can  use fireImmediately
 ```dart
 static final provider = ChangeNotifierProvider((ref) {
    var ret = P2();
    ref.listen<P1>(P1.provider, (previous, next) {
      ret.updateCategories(next.getCategories());
      },
      fireImmediately: true, <--- here
    );
    return ret;
  });
```

**Reply by XuanTung95** [reply]
I wonder if there is a better solution. Or should we request Remi for a new feature?

**Reply by rrousselGit** [reply]
This is the correct solution.

What else would you expect?

**Reply by XuanTung95** [reply]
I expected something more official and easier to use. I think most people don't know how to do this way. And I did not find it in the doc, could you update the doc about this case if you have time?

**Reply by rrousselGit** [reply]
What exactly would "easier to use" look like?

To begin with this scenario should be fairly uncommon.

**Reply by XuanTung95** [reply]
> What exactly would "easier to use" look like?

I think it can look like this

```
class ProviderListen<T,R> {
  final ProviderBase<T> provider;
  final Function(T newState, R currState, Ref ref) onUpdate;

  ProviderListen({required this.provider, required this.onUpdate});
}

final provider = ChangeNotifierProvider(
    (ref) {
      return State1();
    },
    listen: [
      ProviderListen(
        provider: StateCart.provider,
        onUpdate: (newState, currState, ref) {
          currState.update(newState.value);
        },
      ),
    ],
  );
```

> To begin with this scenario should be fairly uncommon.

Why do you think multiple providers depending on each other is uncommon?
This way users can split big provider into smaller one. E.g 1 provider which has to manage ListA, ListB, ListC now will become 3 smaller provider but stil act like the big provider.

**Reply by rrousselGit** [reply]
> I think it can look like this

Why is that simpler? What's the gain over using `ref.listen` as shown before?

> Why do you think multiple providers depending on each other is uncommon?


Depending on other providers is common. Using `ref.listen` to trigger a side-effect on a provider when another provider changes isn't

For example, chances are there's a way to straight-up using `ref.watch`.

**Comment by XuanTung95** [comment]
> Why is that simpler?

Ok, it may not seem more simple. But my point is `ref.listen` can be use for many things, not just to trigger other provider so it confusing whether we should use `ref.listen` or using something else in this case. And it is verbose. Maybe it's not necessary to implement new things.

> Depending on other providers is common.

If it's common, then what people do when 1 provider is changed and others depend on it?

**Reply by XuanTung95** [reply]
@rrousselGit

**Comment by aidanmack** [comment]
Im having a very similar problem.

So basically as I see it, if we watch a provider from a stateNotifier and a change happens on the provider we a watching the stateNotifiers "state" is reset.

For example I have a provider that provides a httpClient. If the user changes the language, then the baseurl has to be recreated. 
So on language change a new httpClient has to be created.


My work around for this is grim. I read the current state of the clientProvider and then I listen for a change.

```
  SignUpStateNotifier({required this.ref}) : super(SignUpNewState()) {
    httpClient = ref.read(httpClientProvider);
    ref.listen(httpClientProvider, (previous, client) {
      onBoardingHttpClient = client;
    });
  }
````

**Reply by rrousselGit** [reply]
That sounds normal. What's the issue with using ref.listen?

Note that there's an optional `fireImmediately:true` parameter to avoid potential duplication

**Reply by aidanmack** [reply]
oh ok - I did not know about `fireImmediately` I dont think I need to read that provider first then. Thats better, thanks.

Maybe there is no issue. I guess the documentation really puts you in the mindset to use `watch` where possible.
But to be honest I suspect I would be using listen quite a bit in my StateNotifers to guard against losing state.

Thanks

**Comment by g-30** [comment]
@rrousselGit imo passing the previous state would be useful for `StateNotifierProvider`.

So instead of this:
```dart
StateNotifierProvider<Notifier, ValueType>((ref) {
  final notifier = Notifier(ValueType());

  ref.listen(anotherProvider, (_, newValue) {
    notifier.updateInnerProperty(newValue);
  }, fireImmediately: true);

  return notifier;
});
```
We could do this:
```dart
StateNotifierProvider<Notifier, ValueType>((ref, ValueType? previousValue) {
  ValueType value = previousValue ?? ValueType();

  final innerProperty = ref.watch(anotherProvider);
  value = value.copyWith(innerProperty: innerProperty);

  return Notifier(value);
});
```

If not, having the `listen(..., fireImmediately: true)` approach defined in the docs would be very helpful. I'm glad I was able to find this thread, wouldn't have figured this one out on my own. Thank you for your work!
